File: supabase/functions/analyze-food/alternatives.ts
import { Alternative, NutritionInfo } from "./types.ts";
import { ensureModelTrained } from "./model.ts";
import { calculateHealthScore, toNumber } from "./scoring.ts";
import { calculateFeatureImportance } from "./explain.ts";
import { getFoodCategory, detectMealType, getCategoryMatchedSearchTerms } from "./classify.ts";
import { searchOpenFoodFacts } from "./openFoodFacts.ts";
import { mealTypeAlternatives, alternativesByCategory } from "./curatedData.ts";

// Normalize name for deduplication (handles variations like "Grilled Chicken" vs "grilled chicken breast")
export function normalizeName(name: string): string {
  return name.toLowerCase()
    .replace(/[^a-z0-9\s]/g, '')
    .replace(/\s+/g, ' ')
    .trim()
    .split(' ')
    .slice(0, 2)  // Take first 2 words for similarity matching
    .join(' ');
}

// Check if a name is similar to any in the seen set
export function isSimilarToSeen(name: string, seenNames: Set<string>): boolean {
  const normalizedNew = normalizeName(name);
  
  for (const seen of seenNames) {
    const normalizedSeen = normalizeName(seen);
    // Check if names share the same first 2 significant words
    if (normalizedNew === normalizedSeen) return true;
    // Also check if one contains the other
    if (normalizedNew.includes(normalizedSeen) || normalizedSeen.includes(normalizedNew)) return true;
  }
  return false;
}

// Curated alternatives as fallback (organized by category AND meal type)
export async function getCuratedAlternatives(
  baselineFood: { identifiedFood: string; nutritionInfo: NutritionInfo }, 
  category: string = 'general', 
  mealType: string = 'any',
  targetBodyType: string = 'athletic',
  baselineScore: number = 0
): Promise<Alternative[]> {
  const name = (baselineFood?.identifiedFood || '').toLowerCase();
  let key = category;

  // Determine specific subcategory
  if (category === 'salty-snack' && name.includes('chip')) {
    key = 'salty-chips';
  } else if (category === 'sweet' && (name.includes('biscuit') || name.includes('cookie'))) {
    key = 'sweet-biscuit';
  } else if (category === 'sweet' && name.includes('chocolate')) {
    key = 'sweet-chocolate';
  }

  // For lunch/dinner items, prefer meal-type specific alternatives
  let alternatives: any[];
  if (mealType === 'lunch-dinner' && mealTypeAlternatives['lunch-dinner']) {
    alternatives = mealTypeAlternatives['lunch-dinner'];
  } else if (mealType === 'breakfast' && (alternativesByCategory['breakfast'] || mealTypeAlternatives['breakfast'])) {
    alternatives = alternativesByCategory['breakfast'] || mealTypeAlternatives['breakfast'];
  } else {
    alternatives = alternativesByCategory[key] || alternativesByCategory[category] || alternativesByCategory.general;
  }

  const model = await ensureModelTrained();
  
  const evaluated = await Promise.all(alternatives.map(async alt => {
    const score = await calculateHealthScore(alt.nutrition, targetBodyType);
    const reasons = calculateFeatureImportance(baselineFood.nutritionInfo, alt.nutrition, model, targetBodyType);

    return {
      name: alt.name,
      healthScore: Math.round(score),
      benefits: [
        `${Math.round(alt.nutrition.protein)}g protein per 100g`,
        `${Math.round(alt.nutrition.fiber)}g fiber per 100g`,
        `${Math.round(alt.nutrition.calories)} calories per 100g`
      ],
      reasons: reasons.length > 0 ? reasons : [
        {
          factor: "Overall Nutrition",
          explanation: "Better nutritional profile for health",
          actualChange: "Improved nutrition",
          status: "better" as const
        }
      ],
      nutrition: alt.nutrition,
      isRegional: false
    };
  }));

  // Only return curated alternatives that actually beat the baseline
  return evaluated.filter(alt => alt.healthScore > baselineScore + 3);
}

// Get healthier alternatives from Open Food Facts & curated data in parallel
export async function getHealthierAlternatives(
  baselineFood: { identifiedFood: string; nutritionInfo: NutritionInfo }, 
  baselineScore: number,
  targetBodyType: string = 'athletic'
): Promise<Alternative[]> {
  const foodName = baselineFood?.identifiedFood || 'snack';
  const category = getFoodCategory(foodName, '');
  const mealType = detectMealType(foodName);
  
  console.log(`Food: ${foodName}, Category: ${category}, Meal Type: ${mealType}, Target: ${targetBodyType}`);
  
  // Generate search terms that match the same category of food
  const searchTerms = getCategoryMatchedSearchTerms(foodName, category, mealType);
  
  const alternatives: Alternative[] = [];
  const seenNames = new Set<string>();
  const model = await ensureModelTrained();

  // Run searches in parallel across top 3 search terms
  const searchResults = await Promise.allSettled(
    searchTerms.slice(0, 3).map(term => searchOpenFoodFacts(term))
  );

  const candidateProducts: any[] = [];
  for (const res of searchResults) {
    if (res.status === 'fulfilled' && Array.isArray(res.value)) {
      candidateProducts.push(...res.value);
    }
  }

  for (const product of candidateProducts) {
    if (!product || !product.nutriments) continue;
    
    const productName = String(product.product_name || '').trim();
    if (!productName || productName.toLowerCase() === 'unknown' || productName.length < 3) continue;
    
    // Strong deduplication - check normalized similarity
    if (isSimilarToSeen(productName, seenNames)) continue;

    // Open Food Facts reports sodium in grams (sodium_100g); convert to milligrams (mg, ×1000)
    const rawSodiumGrams = toNumber(product.nutriments['sodium_100g'] ?? product.nutriments['sodium'], 0);
    const sodiumMg = rawSodiumGrams * 1000;

    const altNutrition: NutritionInfo = {
      calories: toNumber(product.nutriments['energy-kcal_100g'] || (product.nutriments.energy_100g ? product.nutriments.energy_100g / 4.184 : 0), 0),
      protein: toNumber(product.nutriments.proteins_100g, 0),
      carbs: toNumber(product.nutriments.carbohydrates_100g, 0),
      fat: toNumber(product.nutriments.fat_100g, 0),
      saturatedFat: toNumber(product.nutriments['saturated-fat_100g'], 0),
      sugar: toNumber(product.nutriments.sugars_100g, 0),
      fiber: toNumber(product.nutriments.fiber_100g, 0),
      sodium: sodiumMg,
      vitamins: 0.6,
      processingLevel: 0.3
    };

    // Filter out zero-calorie or corrupted empty entries
    if (altNutrition.calories <= 5 && altNutrition.protein <= 0.1 && altNutrition.carbs <= 0.1) continue;

    const altScore = await calculateHealthScore(altNutrition, targetBodyType);

    // Only accept candidates that beat baseline by the margin threshold (+3)
    if (altScore > baselineScore + 3) {
      const reasons = calculateFeatureImportance(baselineFood.nutritionInfo, altNutrition, model, targetBodyType);
      
      if (reasons.length > 0) {
        seenNames.add(productName);
        alternatives.push({
          name: productName,
          healthScore: Math.round(altScore),
          benefits: [
            `${Math.round(altNutrition.protein)}g protein per 100g`,
            `${Math.round(altNutrition.fiber || 0)}g fiber per 100g`,
            `${Math.round(altNutrition.calories)} calories per 100g`
          ],
          reasons: reasons,
          nutrition: altNutrition,
          isRegional: false
        });

        if (alternatives.length >= 3) break;
      }
    }
  }

  // If not enough from API, add curated alternatives matching category ONLY if they beat the baseline
  if (alternatives.length < 3) {
    const curated = await getCuratedAlternatives(baselineFood, category, mealType, targetBodyType, baselineScore);
    const filteredCurated = curated.filter(c => !isSimilarToSeen(c.name, seenNames) && c.healthScore > baselineScore + 3);
    
    for (const alt of filteredCurated) {
      if (alternatives.length >= 3) break;
      if (!isSimilarToSeen(alt.name, seenNames)) {
        seenNames.add(alt.name);
        alternatives.push(alt);
      }
    }
  }

  return alternatives.slice(0, 3);
}

File: python_app/alternatives.py
import re
import asyncio
from typing import List, Set, Dict, Any, Optional
from .models import Alternative, NutritionInfo
from .model import ensure_model_trained
from .scoring import calculate_health_score, to_number
from .explain import calculate_feature_importance
from .classify import get_food_category, detect_meal_type, get_category_matched_search_terms
from .open_food_facts import search_open_food_facts
from .curated_data import meal_type_alternatives, alternatives_by_category

from .allergens import detect_dietary_conflict

def normalize_name(name: str) -> str:
    cleaned = re.sub(r'[^a-z0-9\s]', '', name.lower())
    words = re.sub(r'\s+', ' ', cleaned).strip().split(' ')
    return ' '.join(words[:2])

def is_similar_to_seen(name: str, seen_names: Set[str]) -> bool:
    norm_new = normalize_name(name)
    for seen in seen_names:
        norm_seen = normalize_name(seen)
        if norm_new == norm_seen or norm_new in norm_seen or norm_seen in norm_new:
            return True
    return False

async def get_curated_alternatives(
    baseline_food: Dict[str, Any],
    category: str = 'general',
    meal_type: str = 'any',
    target_body_type: str = 'athletic',
    current_body_type: str = 'average',
    baseline_score: float = 0.0,
    diet_preference: str = 'non-veg'
) -> List[Alternative]:
    name = str(baseline_food.get("identifiedFood", "")).lower()
    key = category

    if category == 'salty-snack' and 'chip' in name:
        key = 'salty-chips'
    elif category == 'sweet' and ('biscuit' in name or 'cookie' in name):
        key = 'sweet-biscuit'
    elif category == 'sweet' and 'chocolate' in name:
        key = 'sweet-chocolate'

    if meal_type == 'lunch-dinner' and 'lunch-dinner' in meal_type_alternatives:
        candidates = meal_type_alternatives['lunch-dinner']
    elif meal_type == 'breakfast' and ('breakfast' in alternatives_by_category or 'breakfast' in meal_type_alternatives):
        candidates = alternatives_by_category.get('breakfast') or meal_type_alternatives.get('breakfast', [])
    else:
        candidates = alternatives_by_category.get(key) or alternatives_by_category.get(category) or alternatives_by_category.get('general', [])

    # Filter candidates by user dietary preference
    filtered_candidates = [
        alt for alt in candidates 
        if detect_dietary_conflict(alt.name, diet_preference) is None
    ]

    model = await ensure_model_trained()
    evaluated = []

    for alt in filtered_candidates:
        score = await calculate_health_score(alt.nutrition, target_body_type, current_body_type)
        reasons = calculate_feature_importance(
            baseline_food["nutritionInfo"],
            alt.nutrition,
            model,
            target_body_type,
            current_body_type
        )
        if score > baseline_score + 2.0:
            evaluated.append(Alternative(
                name=alt.name,
                healthScore=int(round(score)),
                benefits=[
                    f"{int(round(alt.nutrition.protein))}g protein per 100g",
                    f"{int(round(alt.nutrition.fiber or 0))}g fiber per 100g",
                    f"{int(round(alt.nutrition.calories))} calories per 100g"
                ],
                reasons=reasons,
                nutrition=alt.nutrition,
                isRegional=False
            ))

    evaluated.sort(key=lambda a: a.healthScore, reverse=True)
    return evaluated

async def get_healthier_alternatives(
    baseline_food: Dict[str, Any],
    baseline_score: float,
    target_body_type: str = 'athletic',
    current_body_type: str = 'average',
    diet_preference: str = 'non-veg'
) -> List[Alternative]:
    food_name = baseline_food.get("identifiedFood", "snack")
    category = get_food_category(food_name, "")
    meal_type = detect_meal_type(food_name)

    alternatives: List[Alternative] = []
    seen_names: Set[str] = set()
    model = await ensure_model_trained()

    # 1. Fast path: Evaluate curated alternatives in 0ms
    curated = await get_curated_alternatives(
        baseline_food, 
        category, 
        meal_type, 
        target_body_type, 
        current_body_type, 
        baseline_score,
        diet_preference
    )
    for alt in curated:
        if len(alternatives) >= 4:
            break
        if not is_similar_to_seen(alt.name, seen_names):
            seen_names.add(alt.name)
            alternatives.append(alt)

    # 2. Query Open Food Facts if more alternatives needed
    if len(alternatives) < 3:
        search_terms = get_category_matched_search_terms(food_name, category, meal_type)
        tasks = [search_open_food_facts(term) for term in search_terms[:2]]
        search_results = await asyncio.gather(*tasks, return_exceptions=True)

        candidate_products = []
        for res in search_results:
            if isinstance(res, list):
                candidate_products.extend(res)

        for product in candidate_products:
            if len(alternatives) >= 3:
                break
            if not product or not isinstance(product, dict) or not product.get("nutriments"):
                continue

            product_name = str(product.get("product_name", "")).strip()
            if not product_name or product_name.lower() == "unknown" or len(product_name) < 3:
                continue
            if is_similar_to_seen(product_name, seen_names):
                continue
            if detect_dietary_conflict(product_name, diet_preference) is not None:
                continue

            nutriments = product["nutriments"]
            raw_sodium = to_number(nutriments.get("sodium_100g") or nutriments.get("sodium"), 0.0)
            sodium_mg = raw_sodium * 1000.0 if raw_sodium < 20.0 and raw_sodium > 0.0 else raw_sodium

            calories = to_number(
                nutriments.get("energy-kcal_100g") or 
                (float(nutriments.get("energy_100g", 0)) / 4.184 if nutriments.get("energy_100g") else None), 
                0.0
            )
            alt_nutrition = NutritionInfo(
                calories=calories,
                protein=to_number(nutriments.get("proteins_100g"), 0.0),
                carbs=to_number(nutriments.get("carbohydrates_100g"), 0.0),
                fat=to_number(nutriments.get("fat_100g"), 0.0),
                saturatedFat=to_number(nutriments.get("saturated-fat_100g"), 0.0),
                sugar=to_number(nutriments.get("sugars_100g"), 0.0),
                fiber=to_number(nutriments.get("fiber_100g"), 0.0),
                sodium=sodium_mg,
                vitamins=0.6,
                processingLevel=0.3
            )

            if alt_nutrition.calories <= 5 and alt_nutrition.protein <= 0.1 and alt_nutrition.carbs <= 0.1:
                continue

            alt_score = await calculate_health_score(alt_nutrition, target_body_type, current_body_type)
            if alt_score > baseline_score + 2.0:
                reasons = calculate_feature_importance(
                    baseline_food["nutritionInfo"],
                    alt_nutrition,
                    model,
                    target_body_type,
                    current_body_type
                )
                if reasons:
                    seen_names.add(product_name)
                    alternatives.append(Alternative(
                        name=product_name,
                        healthScore=int(round(alt_score)),
                        benefits=[
                            f"{int(round(alt_nutrition.protein))}g protein per 100g",
                            f"{int(round(alt_nutrition.fiber or 0))}g fiber per 100g",
                            f"{int(round(alt_nutrition.calories))} calories per 100g"
                        ],
                        reasons=reasons,
                        nutrition=alt_nutrition,
                        isRegional=False
                    ))

    # Sort and guarantee score differentiation
    alternatives.sort(key=lambda a: a.healthScore, reverse=True)
    final_alts = alternatives[:3]
    for i in range(1, len(final_alts)):
        if final_alts[i].healthScore >= final_alts[i - 1].healthScore:
            final_alts[i].healthScore = max(int(baseline_score) + 1, final_alts[i - 1].healthScore - (i + 1))

    return final_alts
